contribs/analysis/src/main/java/org/matsim/contrib/analysis/vsp/traveltimedistance/GoogleMapRouteValidator.java

```diff
@@ -0,0 +1,119 @@
+import org.apache.logging.log4j.Logger;
+import org.json.simple.JSONArray;
+import org.json.simple.JSONObject;
+import org.json.simple.parser.JSONParser;
+import org.json.simple.parser.ParseException;
+import org.matsim.api.core.v01.Coord;
+import org.matsim.core.utils.collections.Tuple;
+import org.matsim.core.utils.geometry.CoordinateTransformation;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.net.URL;
+import java.text.DecimalFormat;
+import java.text.NumberFormat;
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+import java.time.temporal.ChronoUnit;
+import java.util.Locale;
+
+public class GoogleMapRouteValidator implements TravelTimeDistanceValidator {
+    private static final Logger log = LogManager.getLogger(GoogleMapRouteValidator.class);
+    final String apiAccessKey;
+    final String outputFolder;
+    final String date;
+    final CoordinateTransformation ct;
+    boolean writeDetailedFiles;
+
+    public GoogleMapRouteValidator(String outputFolder, String apiAcessKey, String date, CoordinateTransformation ct, boolean writeDetailedFiles) {
+        this.outputFolder = outputFolder;
+        this.apiAccessKey = apiAcessKey;
+        this.date = date;
+        this.ct = ct;
+        this.writeDetailedFiles = writeDetailedFiles;
+
+        File outDir = new File(outputFolder);
+        if (!outDir.exists()) {
+            outDir.mkdirs();
+        }
+    }
+
+
+    @Override
+    public Tuple<Double, Double> getTravelTime(CarTrip trip) {
+        String tripId = trip.getPersonId().toString() + "_" + trip.getDepartureTime();
+        return getTravelTime(trip.getDepartureLocation(), trip.getArrivalLocation(), trip.getDepartureTime(), tripId);
+    }
+
+    @Override
+    public Tuple<Double, Double> getTravelTime(Coord fromCoord, Coord toCoord, double departureTime, String tripId) {
+        long travelTime = 0;
+        long distance = 0;
+
+        double adjustedDepartureTime = calculateGoogleDepartureTime(departureTime);
+        Coord from = ct.transform(fromCoord);
+        Coord to = ct.transform(toCoord);
+
+        Locale locale = new Locale("en", "UK");
+        String pattern = "###.#######";
+
+        DecimalFormat df = (DecimalFormat) NumberFormat.getNumberInstance(locale);
+        df.applyPattern(pattern);
+
+        String urlString = "https://maps.googleapis.com/maps/api/distancematrix/json" +
+                "?departure_time=" + Long.toString((long) adjustedDepartureTime) +
+                "&destinations=" + df.format(to.getY()) + "%2C" + df.format(to.getX()) +
+                "&origins=" + df.format(from.getY()) + "%2C" + df.format(from.getX()) +
+                "&key=" + apiAccessKey;
+
+        try {
+            System.out.println(urlString);
+            URL url = new URL(urlString);
+            BufferedReader in = new BufferedReader(new InputStreamReader(url.openStream()));
```

Comment:
It looks like this Reader is not closed. Better wrap it in try-with-resource block.

Reply:
Yes.

Reply:
Okay. Thanks!